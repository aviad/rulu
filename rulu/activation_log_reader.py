"""
This module provides a mechanism for tracing the rules activated by the rule
engine, including source and asserted facts for each activation.

Note: facts asserted from Python code are currently not included in the trace.
"""

import clips
import re
from collections import defaultdict

from actions import Assert
from expr import BaseExpr
from ruledef import RuleDef

class ActivationLogReader(object):
    """
    Reader for the rule activation log.
    
    Note: for this to work, the engine must be initialized with trace=True.
    """
    def __init__(self, engine):
        self.engine = engine
        self.environment = engine.environment
        self._read_facts()
        self._read_rules()
        
    def find_fact(self, fact_type, **kwargs):
        """
        Find a fact object by type and field values
        """
        for key, value in kwargs.items():
            if isinstance(value, BaseExpr):
                kwargs[key] = value.evaluate()
        values = tuple(value.evaluate() if isinstance(value, BaseExpr) else value for _, value in 
                       sorted(kwargs.iteritems(), key=lambda x:fact_type._field_order[x[0]]))
        return self.fact_by_content[fact_type._name][values]        

    def get_activations(self):
        """
        Return an iterator over the activation log, including rule object and
        source facts for each activation
        """
        ACTIVATION_RE = re.compile('==> Activation 0\s+([\w@]+): ([f\-\d\,]+)')
        for line in clips.TraceStream.Read().splitlines():
            rule_name, src_fact_ids = ACTIVATION_RE.match(line).groups()
            rule = self.rule_by_name[rule_name]
            src_facts = [self.fact_by_index[int(fact_id.rsplit('-', 1)[1])]
                         for fact_id in src_fact_ids.split(',')]
            yield rule, src_facts
            
    def resolve_asserted_facts(self, rule, src_facts):
        """
        Find the facts that are asserted (using the Assert action) by a given rule 
        and specific source facts. 
        """
        field_map = {}
        fact_map = {}
        premises = rule.premises.values()
        premises.sort(key=lambda p:p.var_name)
        for premise, fact in zip(premises, src_facts):
            premise.container._update_python_param(fact_map, fact)
            for key, value in fact._as_dict().iteritems():
                field_map[getattr(premise.container, key)] = value

        asserted_data = []
        for action in rule.actions:
            if isinstance(action, Assert):
                asserted_data.append(action.replace_fields(field_map).data)
                
        for action in rule.python_actions:
            assert_ = lambda **kwargs: asserted_data.append(kwargs)
            action(assert_=assert_, **fact_map)
            
        for data in asserted_data:
            yield self.find_fact(rule.target, **data)

    def _read_facts(self):
        """ Read and index all facts. """
        self.fact_by_index = {}
        self.fact_by_content = defaultdict(dict)
        for fact in self.engine.get_facts():
            self.fact_by_index[fact._clips_obj.Index] = fact
            self.fact_by_content[fact._name][fact._data] = fact
    
    def _read_rules(self):
        self.rule_by_name = {rule_def._rule.clips_name: rule_def._rule
                             for rule_def in RuleDef._all_instances}
