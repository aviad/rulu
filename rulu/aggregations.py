from bunch import Bunch
from collections import defaultdict

from expr import FieldExpr
from fact import FactReference

class Aggregator(object):
    def __init__(self, engine, template):
        self.engine = engine
        self.template = template
        
    def init(self):
        pass
    
    def finalize(self):
        pass
    
    def process_one(self, **kwargs):
        raise NotImplementedError

    def assert_(self, **kwargs):
        self.engine.assert_(self.template, **kwargs)

def groupby(keys, func):
    keys = tuple(_make_key(key) for key in keys)
    class GroupAggregator(Aggregator):
        def init(self): 
            self.data = defaultdict(lambda : [])
            
        def process_one(self, **kwargs):
            key = tuple(key(**kwargs) for key in keys)
            value = kwargs.values()[0] if len(kwargs) == 1 else Bunch(kwargs)
            self.data[key].append(value)
            
        def finalize(self): 
            for key, group in self.data.iteritems():
                func(group, self.assert_, *key)
        
    return GroupAggregator

def _make_key(key):
    if isinstance(key, FieldExpr):
        return lambda **kwargs : getattr(_make_key(key.container)(**kwargs), key.name)
    elif isinstance(key, FactReference):
        return lambda **kwargs : key._from_python_params(kwargs)
    else:
        raise TypeError('Invalid group key: {}'.format(type(key)))
