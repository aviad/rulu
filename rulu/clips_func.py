from expr import ClipsFuncExpr
from typedefs import Integer

class ClipsFuncProxy(object):
    def __getattr__(self, funcname):
        def func(*args):
            return ClipsFuncExpr(funcname.replace('_', '-'), Integer, *args)
        func.__name__ = funcname
        return func

clips_funcs = ClipsFuncProxy()
