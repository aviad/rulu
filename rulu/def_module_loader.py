import sys
from importlib import import_module
from threading import Lock
from types import ModuleType

from ruledef import RuleDef, params
from func import RuleFunc
from slots import HasSlots
from utils import logger

class DefModuleLoader(object):
    def __init__(self, engine):
        self.engine = engine
        self.lock = Lock()
        self.logger = logger.getChild(type(self).__name__)
        
    def load(self, module_name, package=None, **module_params):
        # Prepare for import
        if isinstance(package, ModuleType):
            package = package.__name__
        if package is not None and '.' not in module_name:
            module_name = '.' + module_name
        self.logger.debug('Loading: %s', module_name)
            
        # Import module, while keeping track of entities that require compilation
        existing_modules = set(sys.modules)
        with self.lock:
            params.update(module_params)
            RuleDef._all_instances = []
            HasSlots._all_subclasses = []
            RuleFunc._set_cur_engine(self.engine)
            module = import_module(module_name, package)
            if module.__name__ in existing_modules:
                reload(module)
            # TODO: Why are there non-named templates??
            to_build = [x for x in HasSlots._all_subclasses if x._name is not None] + RuleDef._all_instances 

        # Add name to all unnamed instances (according to variable name)
        for name in dir(module):
            entity = getattr(module, name)
            if isinstance(entity, RuleDef):
                entity._set_name(name) # TODO: include module name
                
        # Compile everything, except unnamed templates (which cannot be compiled)
        # Note that RuleDefs also may contain TemplateDefs, which would not be named.
        for entity in to_build:
            entity._build(self.engine)
