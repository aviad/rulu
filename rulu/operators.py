from expr import BaseExpr, VariableExpr, normalize_expr
from typedefs import Boolean, Integer, Number
from utils import LispExpr, RuleEngineError

class BaseBinaryOperator(BaseExpr):
    def __init__(self, lhs, rhs, op):
        super(BaseBinaryOperator, self).__init__(all_fields=lhs.all_fields.union(rhs.all_fields))
        self.lhs = lhs
        self.rhs = rhs
        self.op = op
        
    def to_lisp(self):
        return LispExpr(self.op, self.lhs.to_lisp(), self.rhs.to_lisp())
    
    def replace_fields(self, field_map):
        return type(self)(lhs=self.lhs.replace_fields(field_map), 
                          rhs=self.rhs.replace_fields(field_map),
                          op=self.op)
        
    def __str__(self):
        return '({} {} {})'.format(self.lhs, self.op, self.rhs)
    
class ArithmeticBinaryOperator(BaseBinaryOperator, BaseExpr):
    def get_type(self):
        lhs_type, rhs_type = self.lhs.get_type(), self.rhs.get_type()
        if lhs_type in (Integer, Number) and rhs_type in (Integer, Number):
            return Integer if lhs_type is Integer and rhs_type is Integer else Number
        else:
            raise RuleEngineError('Illegal operator "{}" on {} and {}'.format(
                    self.op.__name__, self.lhs.get_type(), self.rhs.get_type()))
    
class BooleanBinaryOperator(BaseBinaryOperator, BaseExpr):
    def get_type(self):
        return Boolean
    
class Condition(BaseExpr):
    def __init__(self, expr):
        expr = normalize_expr(expr)
        super(Condition, self).__init__(all_fields=expr.all_fields)
        if expr.get_type() is not Boolean:
            raise RuleEngineError('Condition must be of boolean type. Got "{}", which is {}'.format(expr, expr.get_type()))
        self.expr = expr
        
    def get_type(self):
        return Boolean

    def to_lisp(self):
        expr = self.expr
        if isinstance(self.expr, VariableExpr):
            expr = not_(not_(expr))
        return LispExpr('test', expr.to_lisp())
    
    def replace_fields(self, field_map):
        return type(self)(self.expr.replace_fields(field_map))
    
    def __str__(self):
        return 'Condition <{}>'.format(self.expr)
    
class UnaryOperator(BaseExpr):
    def __init__(self, expr, op):
        super(UnaryOperator, self).__init__(all_fields=expr.all_fields)
        self.expr = expr
        self.op = op
        
    def get_type(self):
        return self.expr.get_type()
    
    def to_lisp(self):
        return LispExpr(self.op, self.expr.to_lisp())
    
    def replace_fields(self, field_map):
        return type(self)(expr=self.expr.replace_fields(field_map), 
                          op=self.op)
        
    def __str__(self):
        return '({} {})'.format(self.op, self.expr)

or_ = lambda lhs, rhs: BooleanBinaryOperator(normalize_expr(lhs), normalize_expr(rhs), op='or')
not_ = lambda expr: UnaryOperator(normalize_expr(expr), op='not')
