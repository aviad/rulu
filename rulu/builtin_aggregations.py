from actions import Assert, Delete, Update
from clips_func import fact_index, gensym, min_, max_
from expr import normalize_expr
from rule import Rule

UNIQUE_INDEX_FIELD = '_uniqueIndex'

class MapReduceFunc(object):
    def _map(self, x): return x
    def _reduce(self, x, y): raise NotImplementedError
    
    def __init__(self, param):
        self.param = param
        self.map_value = self._map(param)

class AssertAggregate(Assert):
    def __init__(self, **kwargs):
        self.map_reduce_values = {}
        for key, value in kwargs.items():
            if isinstance(value, MapReduceFunc):
                kwargs[key] = value.map_value
                self.map_reduce_values[key] = value
        super(AssertAggregate, self).__init__(**kwargs)
        
    def prepare_rule(self, rule):
        if rule.salience is None:
            rule.salience = -1000
        self.data.update({key : normalize_expr(value) for key, value in rule.groupby.iteritems()})
        self.data[UNIQUE_INDEX_FIELD] = gensym()  
        super(AssertAggregate, self).prepare_rule(rule)
        self._create_reduce_rule(rule)
        
    def _create_reduce_rule(self, map_rule):
        r1 = map_rule.target[0]
        r2 = map_rule.target[1]
        self.reduce_rule = Rule()
        self.reduce_rule.set_name('_Reduce')
        for field_name in map_rule.groupby:
            self.reduce_rule.add_variable(getattr(r1, field_name), getattr(r2, field_name))
        self.reduce_rule.add_condition(fact_index(r1) != fact_index(r2))
        reduce_kwargs = {field_name: value._reduce(getattr(r1, field_name), getattr(r2, field_name))
                         for field_name, value in self.map_reduce_values.iteritems()}
        self.reduce_rule.add_action(Delete(r2))
        self.reduce_rule.add_action(Update(r1, **reduce_kwargs))
        self.reduce_rule.set_salience(map_rule.salience)
        map_rule.add_secondary_rule(self.reduce_rule)
        
class Sum(MapReduceFunc): 
    def _reduce(self, x, y): return x + y

class Max(MapReduceFunc):
    def _reduce(self, x, y): return max_(x, y)
    
class Min(MapReduceFunc):
    def _reduce(self, x, y): return min_(x, y)

class Count(Sum):
    def _map(self, x): return 1
