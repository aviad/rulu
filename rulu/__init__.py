from actions import BaseAction, Assert, Update, Delete
from aggregations import Aggregator, groupby
from class_ import Class, InstanceField
from clips_func import clips_funcs
from expr import (IntegerField, NumberField, StringField, UnicodeField,
                  BooleanField, DateTimeField, TRUE, FALSE, UNKNOWN_BOOL)
from fact import Fact
from func import (RuleFunc, TypedRuleFunc, BooleanRuleFunc, IntegerRuleFunc,
                  NumberRuleFunc, StringRuleFunc)
from operators import or_, not_
from ruledef import (RuleDef, match, action, condition, not_exists, fields,
                     salience, name, description, target, foreach)
from typedefs import (RuleEngineType, Boolean, DateTime, Integer, Number, 
                      String, Unicode) 

__all__ = sorted(x for x in locals() if not x.startswith('_'))
