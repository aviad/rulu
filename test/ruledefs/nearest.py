from rulu import *

class Point(Class):
    X = NumberField()
    Y = NumberField()

class Node(Fact):
    point = InstanceField(Point)
    index = IntegerField()
    
@NumberRuleFunc
def calc_dist(p1, p2):
    return float((p1.X-p2.X)**2 + (p1.Y-p2.Y)**2) ** .5

import nearest_distance_rules
Distance = nearest_distance_rules.make_rule(Node)

Nearest = RuleDef(
    foreach(Distance),
    fields(index1=int, index2=int, point1=Point, point2=Point, distance=float)
)

@Nearest._groupby(Distance.index1)
def find_nearest(distances, assert_, index1):
    assert_(**min(distances, key=lambda d:d.distance)._as_dict()) 
