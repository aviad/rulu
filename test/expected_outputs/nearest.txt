(Node (index 1) (point [p1]))
(Node (index 2) (point [p2]))
(Node (index 3) (point [p3]))
(Node (index 4) (point [p4]))
(Node (index 5) (point [p5]))
(Distance (index1 1) (index2 2) (point1 [p1]) (point2 [p2]) (distance 3.0))
(Distance (index1 1) (index2 3) (point1 [p1]) (point2 [p3]) (distance 5.0))
(Distance (index1 1) (index2 4) (point1 [p1]) (point2 [p4]) (distance 13.0))
(Distance (index1 1) (index2 5) (point1 [p1]) (point2 [p5]) (distance 17.0))
(Distance (index1 2) (index2 1) (point1 [p2]) (point2 [p1]) (distance 3.0))
(Distance (index1 2) (index2 3) (point1 [p2]) (point2 [p3]) (distance 4.0))
(Distance (index1 2) (index2 4) (point1 [p2]) (point2 [p4]) (distance 10.295630140987))
(Distance (index1 2) (index2 5) (point1 [p2]) (point2 [p5]) (distance 15.811388300841896))
(Distance (index1 3) (index2 1) (point1 [p3]) (point2 [p1]) (distance 5.0))
(Distance (index1 3) (index2 2) (point1 [p3]) (point2 [p2]) (distance 4.0))
(Distance (index1 3) (index2 4) (point1 [p3]) (point2 [p4]) (distance 9.055385138137417))
(Distance (index1 3) (index2 5) (point1 [p3]) (point2 [p5]) (distance 12.083045973594572))
(Distance (index1 4) (index2 1) (point1 [p4]) (point2 [p1]) (distance 13.0))
(Distance (index1 4) (index2 2) (point1 [p4]) (point2 [p2]) (distance 10.295630140987))
(Distance (index1 4) (index2 3) (point1 [p4]) (point2 [p3]) (distance 9.055385138137417))
(Distance (index1 4) (index2 5) (point1 [p4]) (point2 [p5]) (distance 10.770329614269007))
(Distance (index1 5) (index2 1) (point1 [p5]) (point2 [p1]) (distance 17.0))
(Distance (index1 5) (index2 2) (point1 [p5]) (point2 [p2]) (distance 15.811388300841896))
(Distance (index1 5) (index2 3) (point1 [p5]) (point2 [p3]) (distance 12.083045973594572))
(Distance (index1 5) (index2 4) (point1 [p5]) (point2 [p4]) (distance 10.770329614269007))
(Nearest (index1 1) (index2 2) (point1 [p1]) (point2 [p2]) (distance 3.0))
(Nearest (index1 2) (index2 1) (point1 [p2]) (point2 [p1]) (distance 3.0))
(Nearest (index1 3) (index2 2) (point1 [p3]) (point2 [p2]) (distance 4.0))
(Nearest (index1 4) (index2 3) (point1 [p4]) (point2 [p3]) (distance 9.055385138137417))
(Nearest (index1 5) (index2 4) (point1 [p5]) (point2 [p4]) (distance 10.770329614269007))
